test/spectests/spectests.cpp

```diff
@@ -0,0 +1,290 @@
+                    if (expected.empty() && !result->stack.empty())
+                    {
+                        fail("Unexpected returned value.");
+                        continue;
+                    }
+
+                    if (result->stack.size() != 1)
+                    {
+                        fail("More than 1 value returned.");
+                        continue;
+                    }
+
+                    const auto expected_type = expected.at(0).at("type").get<std::string>();
+                    uint64_t expected_value;
+                    uint64_t actual_value;
+                    if (expected_type == "i32")
+                    {
+                        expected_value = json_to_value<int32_t>(expected.at(0).at("value"));
+                        actual_value =
+                            static_cast<uint64_t>(static_cast<int32_t>(result->stack[0]));
+                    }
+                    else if (expected_type == "i64")
+                    {
+                        expected_value = json_to_value<int64_t>(expected.at(0).at("value"));
+                        actual_value = result->stack[0];
+                    }
+                    else
+                    {
+                        skip("Unsupported expected type '" + expected_type + "'.");
+                        continue;
+                    }
+
+                    if (expected_value != actual_value)
+                    {
+                        std::stringstream message;
+                        message << "Incorrect returned value. Expected: " << expected_type << " (0x"
+                                << std::hex << expected_value << ") Actual: " << std::dec
+                                << actual_value << " (0x" << std::hex << actual_value << std::dec
+                                << ")";
+                        fail(message.str());
+                        continue;
+                    }
+
+                    pass();
+                }
+                else
+                    skip("Unsupported action type '" + action_type + "'");
+            }
+            else if (type == "assert_trap")
+            {
+                const auto& action = cmd.at("action");
+                const auto action_type = action.at("type").get<std::string>();
+                if (action_type != "invoke")
+                    skip("Unsupported action type '" + action_type + "'");
+
+                auto result = invoke(action);
+                if (!result.has_value())
+                    continue;
+
+                if (!result->trapped)
+                {
+                    fail("Function expected to trap, but it didn't.");
+                    continue;
+                }
+
+                pass();
+            }
+            else if (type == "assert_invalid")
+            {
+                const auto filename = cmd.at("filename").get<std::string>();
+                std::ifstream wasm_file{fs::path{path}.replace_filename(filename)};
```

Comment:
This wasm file loading is done in multiple places so perhaps a helper would be due: `bytes_view load_wasm_file(filename)`?

Reply:
Maybe even including the parser stage?